### ISSUE TITLE
Is it possible to add http EventSource to get streaming events from a server?

### ISSUE THREAD

**Issue by Madd0g** [Hammerspoon/hammerspoon#3442]
I have a local server that only supports EventSource and I'd love to plug the output into HS, is that possible with the http client or could it be added?

Thanks

**Comment by Rhys-T** [comment]
It doesn't look like `hs.http` currently has any support for server-sent events (a.k.a. EventSource/`text/event-stream`). All its functions seem to be built to retrieve an entire file/resource at once - there's no way to get any sort of 'streaming' response from it. Looking at the underlying Cocoa API, though, it looks like this could probably be done - perhaps an `hs.http.doAsyncStreamingRequest(...)` function, with an `hs.http.eventsource` module built on top of it? (There are also libraries for using a `text/event-stream` from a Cocoa app, but [the only one I've found so far that's been updated within the last year][LDSwiftEventSource] is written in Swift and uses the Apache-2.0 license, and I don't know what Hammerspoon's policy on either of those things is.)

[LDSwiftEventSource]: https://github.com/launchdarkly/swift-eventsource

For now, here's the options I see for watching an event stream like that from Hammerspoon:
- Use an [`hs.task`][task] to run `/usr/bin/curl -s 'http://some-local-server.example/events'`, and use [setStreamingCallback][ssc] to watch its output and parse it into events.
- Use [`hs.socket`][socket] to speak HTTP 'directly', and parse the events in its callback. Might be doable if you're okay with using HTTP/1.x, but probably way too complicated if you need HTTP/2 or HTTP/3.
- I also considered the possibility of creating a hidden [`hs.webview`][webview] whose only job was to watch the events using the JavaScript EventSource API and relay them back to Hammerspoon through a [user content controller][usercontent]. However, depending on how the server is configured, there's a good chance you'll run into cross-origin restrictions with that approach - especially if the fake 'web page' isn't running from an actual `http[s]://` URL.

[task]: https://www.hammerspoon.org/docs/hs.task.html
[ssc]: https://www.hammerspoon.org/docs/hs.task.html#setStreamingCallback
[socket]: https://www.hammerspoon.org/docs/hs.socket.html
[webview]: https://www.hammerspoon.org/docs/hs.webview.html
[usercontent]: https://www.hammerspoon.org/docs/hs.webview.usercontent.html

**Comment by Madd0g** [comment]
Appreciate the research, thanks!

I might try the `hs.task` approach, seems like the easiest.

**Comment by Rhys-T** [comment]
Ah - I tried `curl` from the terminal to see if it could do streaming output, but didn't think to actually try it from `hs.task`. Looks like it works if I add the `-N` (`--no-buffer`) option, so that `curl` sends the data out to Hammerspoon immediately when it receives it. My guess is that `curl` turns that on by default when it sees that it's writing to a terminal, but not when it's writing to a pipe like it is here.

Forgot to mention before: the `-s` (`--silent`) option I included before is to make sure that `curl` doesn't try to show its progress on `stderr`. I just checked, and that does seem to still be necessary with `hs.task`, if you don't want the callback constantly being called with the progress-meter output.

One more thing: remember that there's no guarantee that the callback will get exactly one message per call. Depending on how exactly TCP (or the pipe coming out of `curl`) decides to break things up, it could be multiple messages, or a message could be split across multiple calls, or both.

**Comment by Madd0g** [comment]
oh nice, adding the `-N` param did the trick! Thanks, now to work on the fun parts!

> or a message could be split across multiple calls

I hope that doesn't happen, they're json

**Comment by Rhys-T** [comment]
Yeah, TCP doesn't really have a concept of discrete 'messages' - the data's all just one big stream of bytes as far as it's concerned, so it can chop it up however it wants to (and the same goes for Unix pipes, and whatever HTTP/2 and HTTP/3 are doing). So anything running on top has to handle that when it's receiving messages (i.e. events).

I haven't done very extensive testing on the following code (I don't have much to test against that uses EventSource), but in theory it should correctly parse a `text/event-stream` according to [the spec][], and pass each event to the `onEvent` function as a table with `type` and `data` fields. (I don't know whether the server you're trying to talk to uses different event types or not, but I figured I'd include it just in case.) It doesn't currently handle event IDs or automatically reconnecting to the server, but that can be added if necessary.

[the spec]: https://html.spec.whatwg.org/multipage/server-sent-events.html#event-stream-interpretation

```lua
local function onTerminate()
	-- whatever you have in onTerminate now, I guess
end
local function onEvent(event)
	local eventType = event.type
	local eventData = hs.json.decode(event.data)
	-- do something
	print(eventType, hs.inspect(eventData))
end

local function parseEvent(eventText)
	local event = {
		type = '',
		data = '',
	}
	for line in eventText:gmatch('[^\n]+') do
		local name, value = line:match('^([^:]*): ?(.*)$')
		if not name then
			name, value = line, ''
		end
		if name == 'event' then
			event.type = value
		elseif name == 'data' then
			event.data = event.data .. value .. '\n'
		else
			-- not currently handling last event ID / auto-reconnection, so ignore `id`/`retry`
			-- any other field is supposed to be ignored anyway
		end
	end
	if event.type == '' then
		event.type = 'message'
	end
	return event
end

local partialEventData = ''
local lastChunkEndedInCR = false
local function onStreamingMsg(task, output, errorOutput)
	if lastChunkEndedInCR and output:sub(1,1) == '\n' then
		-- stopped in the middle of a CRLF
		-- already changed the CR to an LF below, so discard the new LF
		output = output:sub(2)
	end
	lastChunkEndedInCR = output:sub(-1) == '\r'
	output = partialEventData .. output
	-- Normalize any of the supported line endings (CR, LF, CRLF) to LF
	output = output:gsub('\r\n', '\n'):gsub('\r', '\n')
	local startPos = 0
	while true do
		local eventText, endPos = output:match('(.-)\n\n()', startPos)
		if eventText then
			onEvent(parseEvent(eventText))
			startPos = endPos
		else
			break
		end
	end
	partialEventData = output:sub(startPos)
	return true
end

hotelWatcher = hs.task.new('/usr/bin/curl', onTerminate, onStreamingMsg, {'-Ns', 'http://localhost:2000/_/events/output'}):start()
```

**Comment by Madd0g** [comment]
Wow that's great, thanks, I'll give it a whirl.

I see hammerspoon even terminates the task on restart, thoughtful. Wish I'd checked before writing logic to kill the old curl on restart.
